# Importing libraries

In [1]:
import numpy as np
import tensorflow as tf

# Loading data

In [2]:
npz = np.load('AB_train.npz')
train_inputs = npz['inputs'].astype(np.float)
train_targets = npz['targets'].astype(np.int)

npz = np.load('AB_validation.npz')
validation_inputs = npz['inputs'].astype(np.float)
validation_targets = npz['targets'].astype(np.int)

npz = np.load('AB_test.npz')
test_inputs = npz['inputs'].astype(np.float)
test_targets = npz['targets'].astype(np.int)

# Building model and training

In [3]:
input_size = 10
output_size = 2
hidden_layer_size = 50 # Need to experiment and decide value
patience = 2

early_stopping = tf.keras.callbacks.EarlyStopping(patience=patience) # This will monitor validation loss and stop training the first time validation loss increases

model = tf.keras.Sequential([
    tf.keras.layers.Dense(hidden_layer_size, activation='relu'),
    tf.keras.layers.Dense(hidden_layer_size, activation='relu'),
    tf.keras.layers.Dense(output_size, activation='softmax')
])

model.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics=['accuracy']) # sparse_categorical_crossentropy for one hot encoding

batch_size = 100
max_epochs = 100

model.fit(train_inputs, train_targets, batch_size=batch_size, epochs=max_epochs,
         validation_data=(validation_inputs, validation_targets), 
          callbacks=[early_stopping],
          verbose=2)

Train on 3579 samples, validate on 447 samples
Epoch 1/100
3579/3579 - 1s - loss: 0.5404 - accuracy: 0.8005 - val_loss: 0.4296 - val_accuracy: 0.8546
Epoch 2/100
3579/3579 - 0s - loss: 0.3620 - accuracy: 0.8790 - val_loss: 0.3632 - val_accuracy: 0.8658
Epoch 3/100
3579/3579 - 0s - loss: 0.3138 - accuracy: 0.8860 - val_loss: 0.3406 - val_accuracy: 0.8725
Epoch 4/100
3579/3579 - 0s - loss: 0.2935 - accuracy: 0.8947 - val_loss: 0.3203 - val_accuracy: 0.8792
Epoch 5/100
3579/3579 - 0s - loss: 0.2798 - accuracy: 0.8994 - val_loss: 0.3104 - val_accuracy: 0.8792
Epoch 6/100
3579/3579 - 0s - loss: 0.2707 - accuracy: 0.9003 - val_loss: 0.2979 - val_accuracy: 0.8859
Epoch 7/100
3579/3579 - 0s - loss: 0.2624 - accuracy: 0.9028 - val_loss: 0.2927 - val_accuracy: 0.8904
Epoch 8/100
3579/3579 - 0s - loss: 0.2556 - accuracy: 0.9053 - val_loss: 0.2852 - val_accuracy: 0.8949
Epoch 9/100
3579/3579 - 0s - loss: 0.2514 - accuracy: 0.9092 - val_loss: 0.2869 - val_accuracy: 0.8949
Epoch 10/100
3579/3579 - 0

We've managed to get a validation accuracy of 90%.

# Testing model

In [4]:
test_loss, test_accuracy = model.evaluate(test_inputs, test_targets)
print(test_loss, test_accuracy)

448/448 [==============================] - 0s 403us/sample - loss: 0.2491 - accuracy: 0.8951
0.24909172632864543 0.89508927


Testing accuracy is 89%. 

Once tested we are no longer allowed to train the same model before re-compiling. This is because our model has now seen the test dataset.